### PR TITLE
WIP: CAPI AWS implementation

### PR DESCRIPTION
**Pull request by patrickdillon**
This is a WIP implementation of installing AWS using CAPI providers and the interface from #7824. The intent is to provide a *reference* for installing a complete, production-level cluster.

Depends on #7824 
(the first four commits are from that PR)

### CODE REVIEW

File: pkg/asset/cluster/metadata/metadata.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package metadata
+
+import (
+	"encoding/json"
+	"os"
+	"path/filepath"
+
+	"github.com/openshift/installer/pkg/types"
+	"github.com/pkg/errors"
```
Comment (r4f4) [inline pkg/asset/cluster/metadata/metadata.go:9]: Let's not introduce this lib in new code.

File: pkg/asset/cluster/metadata/metadata.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+func Load(dir string) (*types.ClusterMetadata, error) {
+	path := filepath.Join(dir, FileName)
+	raw, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	var metadata *types.ClusterMetadata
+	if err = json.Unmarshal(raw, &metadata); err != nil {
+		return nil, errors.Wrapf(err, "failed to Unmarshal data from %q to types.ClusterMetadata", path)
```
Comment (r4f4) [inline pkg/asset/cluster/metadata/metadata.go:26]:
```suggestion
		return nil, fmt.Errorf("failed to Unmarshal data from %q to types.ClusterMetadata: %w", path, err)
```

File: pkg/infrastructure/aws/clusterapi/clusterapi.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/elbv2"
+	"github.com/aws/aws-sdk-go/service/route53"
+	"github.com/pkg/errors"
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/clusterapi.go:11]: Ditto.

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/pkg/errors"
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:12]: Ditto

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+)
+
+// putIAMRoles creates the roles used by control-plane and compute nodes.
+func putIAMRoles(clusterID string, ic *installconfig.InstallConfig) error {
+	logrus.Infoln("Creating IAM roles for control-plane and compute nodes")
+	// Create the IAM Role with the aws sdk.
+	// https://docs.aws.amazon.com/sdk-for-go/api/service/iam/#IAM.CreateRole
+	session, err := ic.AWS.Session(context.TODO())
+	if err != nil {
+		return errors.Wrap(err, "failed to load AWS session")
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:95]:
```suggestion
		return fmt.Errorf("failed to load AWS session: %w", err)
```

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+				},
+				Action: iamv1.Actions{
+					"sts:AssumeRole",
+				},
+			},
+		},
+	}
+	assumePolicyBytes, err := json.Marshal(assumePolicy)
+	if err != nil {
+		return errors.Wrap(err, "failed to marshal assume policy")
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:122]:
```suggestion
		return fmt.Errorf("failed to marshal assume policy: %w", err)
```

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+	}
+
+	for _, role := range []string{"master", "worker"} {
+		roleName := aws.String(fmt.Sprintf("%s-%s-role", clusterID, role))
+		if _, err := svc.GetRole(&iam.GetRoleInput{RoleName: roleName}); err != nil {
+			if aerr, ok := err.(awserr.Error); ok && aerr.Code() != iam.ErrCodeNoSuchEntityException {
+				return errors.Wrapf(err, "failed to get %s role", role)
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:129]:
```suggestion
				return fmt.Errorf("failed to get %s role: %w", role, err)
```

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+	}
+	assumePolicyBytes, err := json.Marshal(assumePolicy)
+	if err != nil {
+		return errors.Wrap(err, "failed to marshal assume policy")
+	}
+
+	for _, role := range []string{"master", "worker"} {
+		roleName := aws.String(fmt.Sprintf("%s-%s-role", clusterID, role))
+		if _, err := svc.GetRole(&iam.GetRoleInput{RoleName: roleName}); err != nil {
+			if aerr, ok := err.(awserr.Error); ok && aerr.Code() != iam.ErrCodeNoSuchEntityException {
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:128]: Should use `errors.As`

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+				return errors.Wrapf(err, "failed to get %s role", role)
+			}
+			// If the role does not exist, create it.
+			createRoleInput := &iam.CreateRoleInput{
+				RoleName:                 roleName,
+				AssumeRolePolicyDocument: aws.String(string(assumePolicyBytes)),
+				Tags:                     []*iam.Tag{clusterOwnedIAMTag},
+			}
+			if _, err := svc.CreateRole(createRoleInput); err != nil {
+				return errors.Wrapf(err, "failed to create %s role", role)
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:138]:
```suggestion
				return errors.Errorf("failed to create %s role: %w", role, err)
```

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+			}
+			time.Sleep(10 * time.Second)
+			if err := svc.WaitUntilRoleExists(&iam.GetRoleInput{RoleName: roleName}); err != nil {
+				return errors.Wrapf(err, "failed to wait for %s role to exist", role)
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:142]:
```suggestion
				return fmt.Errorf("failed to wait for %s role to exist: %w", role, err)
```

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+			}
+		}
+
+		// Put the policy inline.
+		policyName := aws.String(fmt.Sprintf("%s-%s-policy", clusterID, role))
+		b, err := json.Marshal(policies[role])
+		if err != nil {
+			return errors.Wrapf(err, "failed to marshal %s policy", role)
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:150]:
```suggestion
			return fmt.Errorf("failed to marshal %s policy: %w", role, err)
```

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+		}
+		if _, err := svc.PutRolePolicy(&iam.PutRolePolicyInput{
+			PolicyDocument: aws.String(string(b)),
+			PolicyName:     policyName,
+			RoleName:       roleName,
+		}); err != nil {
+			return errors.Wrapf(err, "failed to create inline policy for role %s ", role)
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:157]:
```suggestion
			return fmt.Errorf("failed to create inline policy for role %s: %w", role, err)
```

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+		}
+
+		profileName := aws.String(fmt.Sprintf("%s-%s-profile", clusterID, role))
+		if _, err := svc.GetInstanceProfile(&iam.GetInstanceProfileInput{InstanceProfileName: profileName}); err != nil {
+			if aerr, ok := err.(awserr.Error); ok && aerr.Code() != iam.ErrCodeNoSuchEntityException {
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:162]: Should use `errors.As`

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+				return errors.Wrapf(err, "failed to get %s instance profile", role)
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:163]:
```suggestion
				return fmt.Errorf("failed to get %s instance profile: %w", role, err)
```

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+			}
+			// If the profile does not exist, create it.
+			if _, err := svc.CreateInstanceProfile(&iam.CreateInstanceProfileInput{
+				InstanceProfileName: profileName,
+				Tags:                []*iam.Tag{clusterOwnedIAMTag},
+			}); err != nil {
+				return errors.Wrapf(err, "failed to create %s instance profile", role)
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:170]:
```suggestion
				return fmt.Errorf("failed to create %s instance profile: %w", role, err)
```

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+			}
+			time.Sleep(10 * time.Second)
+			if err := svc.WaitUntilInstanceProfileExists(&iam.GetInstanceProfileInput{InstanceProfileName: profileName}); err != nil {
+				return errors.Wrapf(err, "failed to wait for %s role to exist", role)
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:174]:
```suggestion
				return fmt.Errorf("failed to wait for %s role to exist: %w", role, err)
```

File: pkg/infrastructure/aws/clusterapi/iam.go [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+			}
+
+			// Finally, attach the role to the profile.
+			if _, err := svc.AddRoleToInstanceProfile(&iam.AddRoleToInstanceProfileInput{
+				InstanceProfileName: profileName,
+				RoleName:            roleName,
+			}); err != nil {
+				return errors.Wrapf(err, "failed to add %s role to instance profile", role)
```
Comment (r4f4) [inline pkg/infrastructure/aws/clusterapi/iam.go:182]:
```suggestion
				return fmt.Errorf("failed to add %s role to instance profile: %w", role, err)
```

File: pkg/infrastructure/clusterapi/clusterapi.go [hunk 18]
```diff
@@ -0,0 +1,279 @@
+package clusterapi
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
```
Comment (r4f4) [inline pkg/infrastructure/clusterapi/clusterapi.go:8]: Ditto

File: pkg/infrastructure/clusterapi/clusterapi.go [hunk 19]
```diff
@@ -0,0 +1,279 @@
...
+	for _, m := range capiManifestsAsset.RuntimeFiles() {
+		manifests = append(manifests, m.Object)
+	}
+
+	// Gather the ignition files, store them in a secret, and add them to manifests.
+	{
+		masterIgn := string(masterIgnAsset.Files()[0].Data)
+		bootstrapIgn, err := injectInstallInfo(bootstrapIgnAsset.Files()[0].Data)
+		if err != nil {
+			return fileList, errors.Wrap(err, "unable to inject installation info")
```
Comment (r4f4) [inline pkg/infrastructure/clusterapi/clusterapi.go:97]:
```suggestion
			return fileList, fmt.Errorf("unable to inject installation info: %w", err)
```

File: pkg/infrastructure/clusterapi/clusterapi.go [hunk 20]
```diff
@@ -0,0 +1,279 @@
...
+
+		gvk, err := cl.GroupVersionKindFor(m)
+		if err != nil {
+			return fileList, fmt.Errorf("failed to get GVK for manifest: %w", err)
+		}
+		fileName := fmt.Sprintf("%s-%s-%s.yaml", gvk.Kind, m.GetNamespace(), m.GetName())
+		objData, err := yaml.Marshal(m)
+		if err != nil {
+			errMsg := fmt.Sprintf("failed to create infrastructure manifest %s from InstallConfig", fileName)
+			return fileList, errors.Wrapf(err, errMsg)
```
Comment (r4f4) [inline pkg/infrastructure/clusterapi/clusterapi.go:243]:
```suggestion
			return fileList, fmt.Errorf("failed to create infrastructure manifest %s from InstallConfig: %w", fileName, err)
```

**Comment by patrickdillon** [comment]
Switched the config so that CAPI creates the int LB (and use SDK to create ext) in ecb455131e65f3027669112b1462c1cc38a5273e

Now installs are completing!

**Pull request closed without merging**